pkg/sessiontxn/isolation/base.go

```diff
@@ -268,6 +270,10 @@ func (p *baseTxnContextProvider) getTxnStartTS() (uint64, error) {
 	return txn.StartTS(), nil
 }
 
+func (p *baseTxnContextProvider) usePresetStartTS() bool {
+	return p.constStartTS != 0 || p.sctx.GetSessionVars().SnapshotTS != 0
```

Comment:
I'm not sure about this: `SnapshotTS != 0` should imply a staleness txn ctx provider? It won't hurt though

Reply:
I am not sure either. That's why I named it `Preset`. The point is we don't check StartTS if it is preset.